README.md
The Logstash Stack
==================

This cookbook includes recipes to set up a Logstash
(http://www.logstash.net) installation.

Requirements
------------

### Packages

This cookbook relies on system packages for the software used. The
packages can be built using a Rake script in the `packages/`
subdirectory:

    $ bundle install
    $ bundle exec rake

These packages should be either preinstalled by another cookbook, or
available in an apt repository.

### Cookbooks

 * apache2
 * apt
 * ark
 * java
 * runit

Attributes
----------

 * `node['logstash_stack']['elasticsearch']['config'][…]` -- `elasticsearch.yml` configuration 
 * `node['logstash_stack']['redis']['use_distribution_packages']` (default: `false`) -- use distribution's `redis-server` package rather than a PPA (meaningful only on Ubuntu)

 * `node['logstash']['default']['java_opts']` (default: `'-server -Djava.awt.headless=true -Xmx512M -Xms128M'`) -- default agent's Java options

 * `node['logstash'][agent_name][…]` -- configuration for a given Logstash intance:
   * `node['logstash'][agent_name][…]['java_opts']` (default: `node['logstash']['default']['java_opts']`) -- Java options for the agent
   * `node['logstash'][agent_name]['input'][…]` -- agent's inputs (see the *Usage* section)
   * `node['logstash'][agent_name]['filter'][…]` -- agent's filters (see the *Usage* section)
   * `node['logstash'][agent_name]['output'][…]` -- agent's outputs (see the *Usage* section)

 * `node['logstash_stack']['kibana3']['domain']` (default: `node.fqdn`) -- domain to serve Kibana at
 * `node['logstash_stack']['kibana3']['apache2_custom']` (default: `nil`) -- custom snippet of Apache config for Kibana (e.g. authentication)
 * `node['logstash_stack']['kibana3']['ssl_key_path']`,
   `node['logstash_stack']['kibana3']['ssl_certificate_path']`,
   `node['logstash_stack']['kibana3']['ssl_certificate_chain_path']` -- if set, Kibana will be served over https
 * `node['logstash_stack']['kibana3']['elasticsearch']` (default: `'http://127.0.0.1:9200/'`) -- URL at which Apache serving Kibana can access Logstash's Elasticsearch

Definitions
-----------

`logstash_agent` -- sets up a Logstash agent with given name

Recipes
-------

### logstash_stack::default

The default use case: sets up Elasticsearch, Redis, `indexer` Logstash
agent, and Kibana frontend.

### logstash_stack::elasticsearch, logstash_stack::redis

These recipes install Elasticsearch for Logstash output, and Redis for
Logstash input.

### logstash_stack::indexer

Installs Logstash agent named `indexer`, configured by default to
take input from Redis, syslog TCP, and Lumberjack, and output to
Elasticsearch.

### logstash_stack::kibana

Installs Kibana 3 served by Apache

### logstash_stack::shipper

Configures machine as a Logstash shipper that ships to the indexer
instance (one that runs `logstash_stack::indexer` recipe).

If the indexer is configured with syslog input, and
`node['logstash']['shipper']['rsyslog']` is true (default), then
rsyslog is configured to ship all events to the Logstash indexer.

If the indexer is configured with lumberjack input, and there are
entries configured in `node['logstash']['shipper']['lumberjack']`,
then Lumberjack is installed and configured to send events to the
indexer.

Finally, if any input is configured in
`node['logstash']['shipper']['input']`, then a Logstash agent named
`shipper` is configured to ship events to the indexer using the
`lumberjack` protocol.

The templates used in shipper's config can use `@params[:indexer]` to
access indexer's Node instance.

Logstash Agent Configuration
-------------------------------

The main piece of configuration is configuring the Logstash agents;
either manually by calling out to the `logstash_agent` definition,
or by using the `logstash_stack::indexer` or `logstash_stack` recipe
which call `logstash_agent "indexer"`. The agent definition
usually takes just a name, and its config file is generated from node's
attributes.

The `node['logstash'][agent_name]['input'][…]`,
`node['logstash'][agent_name]['filter'][…]`, and
`node['logstash'][agent_name]['output'][…]` are dictionaries that
will be used to compile agent configuration's input, filter, and
output sections. A section is written in a following way:

1. Take all entries, order them by their keys
2. If the value is `nil` or `false`, skip the entry - this can be used
   to skip a section that's enabled by default
3. If the value is a string, paste it verbatim
4. If the value is a hash, render it as a subtemplate; the `template`
   will be template name, and rest of the hash will be passed as a
   subtemplate's options. Additionally, subtemplate will get variables
   `@agent` with the agent name, and `@params` with parameters
   of the `logstash_agent` call.

The subtemplate setup may not be clear; here are examples from the
default `indexer` agent's configuration:

```ruby
default['logstash']['indexer']['input']['50-redis']['template'] = 'logstash-input-redis.conf.erb'
default['logstash']['indexer']['input']['50-redis']['cookbook'] = 'logstash-stack'
default['logstash']['indexer']['input']['50-redis']['variables']['host'] = '127.0.0.1'
```

The corresponding `logstash-input-redis.conf.erb` template look like this:

```
redis {
  host => "<%= @host %>"
  type => "redis-input"
  data_type => "list"
  key => "logstash"
  codec => json
}

License and Authors
-------------------
Authors: Maciej Pasternacki <maciej@3ofcoins.net>

License: MIT (see the LICENSE file for details)


CHANGELOG.md
CHANGELOG for logstash
======================

This file is used to list changes made in each version of logstash.

0.4.0
-----

* Renamed `logstash_instance` to `logstash_agent`
* Use dictionary-based single-file instance configuration rather than including
  directory; single file configuration is more
  predictable. `logstash_config` definition has been obviously
  deleted.
* Updated documentation
* Added separate LICENSE and CONTRIBUTING.md files

0.3.0
-----

* Bump to logstash 1.2.2 / elasticsearch 0.90.3
* Bring back packages/ dir
* Kibana 3 proxied by Apache

0.2.0
-----

* Use Kibana rather than logstash-web
* Move package Rakefile to root

0.1.0
-----

* Initial release of logstash

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


attributes/default.rb
default['logstash_stack']['elasticsearch']['config']['cluster.name'] = 'logstash'
default['logstash_stack']['elasticsearch']['config']['node.name'] = name
default['logstash_stack']['elasticsearch']['config']['node.data'] = true

default['logstash_stack']['redis']['use_distribution_packages'] = false

default['logstash_stack']['kibana3']['version'] = '3.0.0milestone4'
default['logstash_stack']['kibana3']['url'] = "https://download.elasticsearch.org/kibana/kibana/kibana-#{node['logstash_stack']['kibana3']['version']}.tar.gz"
default['logstash_stack']['kibana3']['checksum'] = '3ebaac69439aa1925c7918e008978b8424840f3bd3910379d4f2bcf5fdfd2118'
default['logstash_stack']['kibana3']['elasticsearch'] = 'http://127.0.0.1:9200/'
default['logstash_stack']['kibana3']['domain'] = fqdn
default['logstash_stack']['kibana3']['apache2_custom'] = nil

default['logstash']['default']['java_opts'] = '-server -Djava.awt.headless=true -Xmx512M -Xms128M'

default['logstash']['indexer']['input']['50-redis']['template'] = 'logstash-input-redis.conf.erb'
default['logstash']['indexer']['input']['50-redis']['cookbook'] = 'logstash-stack'
default['logstash']['indexer']['input']['50-redis']['variables']['host'] = '127.0.0.1'

default['logstash']['indexer']['input']['50-lumberjack']['template'] = 'logstash-input-lumberjack.conf.erb'
default['logstash']['indexer']['input']['50-lumberjack']['cookbook'] = 'logstash-stack'
default['logstash']['indexer']['input']['50-lumberjack']['variables']['port'] = 5043

default['logstash']['indexer']['input']['50-syslog']['template'] = 'logstash-input-syslog.conf.erb'
default['logstash']['indexer']['input']['50-syslog']['cookbook'] = 'logstash-stack'
default['logstash']['indexer']['input']['50-syslog']['variables']['port'] = 4514

default['logstash']['indexer']['filter']['50-syslog']['template'] = 'logstash-filter-syslog.conf.erb'
default['logstash']['indexer']['filter']['50-syslog']['cookbook'] = 'logstash-stack'

default['logstash']['indexer']['output']['50-elasticsearch']['template'] = 'logstash-output-elasticsearch.conf.erb'
default['logstash']['indexer']['output']['50-elasticsearch']['cookbook'] = 'logstash-stack'
default['logstash']['indexer']['output']['50-elasticsearch']['variables']['host'] = '127.0.0.1'

default['logstash']['shipper']['java_opts'] = '-server -Djava.awt.headless=true -Xmx32M -Xms4M'
default['logstash']['shipper']['rsyslog'] = true
default['logstash']['shipper']['lumberjack'] = {}

default['logstash']['shipper']['output']['50-lumberjack']['template'] = 'logstash-output-lumberjack.conf.erb'
default['logstash']['shipper']['output']['50-lumberjack']['cookbook'] = 'logstash-stack'

